File: CPP/ChangeDynamic.cpp
#include <iostream>
#include <vector>

using namespace std;

//O(amount * num_coins)
vector<int> changeDynamic(int amount, int coins[], int num_coins) {
    vector<int> min_coins(amount + 1, amount + 2);
    vector<int> last_coin_used(amount + 1, -1);

    min_coins[0] = 0;
    last_coin_used[0] = 0;

    for (int i = 1; i <= amount; i++) {
        for (int j = 0; j < num_coins; j++) {
            if (coins[j] <= i && min_coins[i - coins[j]] + 1 < min_coins[i]) {
                min_coins[i] = min_coins[i - coins[j]] + 1;
                last_coin_used[i] = j;
            }
        }
    }

    vector<int> change;
    if (min_coins[amount] == amount + 2) {
        return change;
    }

    while (amount > 0) {
        change.push_back(coins[last_coin_used[amount]]);
        amount -= coins[last_coin_used[amount]];
    }

    return change;
}

int main() {
    int denominations[] = {20, 19, 1};
    int size_of_array = sizeof(denominations) / sizeof(denominations[0]);
    vector<int> result = changeDynamic(58, denominations, size_of_array);
    for(int denomination : result){
        cout << denomination << ", ";
    }
}

File: Python/ChangeDynamic.py
#O(amount * len(denominations))
def change_dynamic(amount: int, denominations: list) -> list:
    min_coins = [amount + 2] * (amount + 1)
    last_denom = [-1] * (amount + 1)
    
    min_coins[0] = 0
    last_denom[0] = 0
    
    for i in range(1, amount + 1):
        for j, denom in enumerate(denominations):
            if denom <= i and min_coins[i - denom] + 1 < min_coins[i]:
                min_coins[i] = min_coins[i - denom] + 1
                last_denom[i] = j

    change = []
    while amount > 0:
        if last_denom[amount] == -1:
            return []
        change.append(denominations[last_denom[amount]])
        amount -= denominations[last_denom[amount]]

    return change

print(change_dynamic(58, [20, 19, 1]))
